File: processing/syntactic/expressions/atomic/await_expr.py
from __future__ import annotations

from processing.codegen.code_builder import CodeBuilder
from processing.lexical.tokens.token import Token
from processing.lexical.tokens.token_type import TokenType
from processing.syntactic.expressions.expr import Expr, child_property
from processing.syntactic.expressions.groups import StatementExpression, AtomExpression
from processing.location import span_marker


class AwaitExpr(AtomExpression, StatementExpression):
    """ await_expr:
        'await' expr_list;
    """

    @child_property
    def value(self) -> Expr: pass

    def __init__(
            self,
            parent: Expr = None,
            await_token: Token = None,
            value: Expr = None
    ):
        super().__init__(parent)
        self.await_token = await_token
        self.value = value

    @span_marker
    def parse(self) -> AwaitExpr:
        self.await_token = self.stream.eat(TokenType.keyword_await)
        self.value = self.parse_multiple(self.parse_any)
        return self

    def to_axion(self, c: CodeBuilder):
        c += self.await_token, self.value

    def to_csharp(self, c: CodeBuilder):
        c += 'await (', self.value, ')'

    def to_python(self, c: CodeBuilder):
        c += 'await ', self.value

File: Axion.Core/Processing/Syntactic/Expressions/Atomic/AwaitExpr.cs
using Axion.Core.Processing.Lexical.Tokens;
using Axion.Core.Processing.Syntactic.Expressions.Common;
using Axion.Core.Processing.Syntactic.Expressions.Generic;
using Axion.Core.Processing.Syntactic.Expressions.TypeNames;
using static Axion.Core.Processing.Lexical.Tokens.TokenType;

namespace Axion.Core.Processing.Syntactic.Expressions.Atomic {
    /// <summary>
    ///     <c>
    ///         await-expr:
    ///             'await' multiple-expr;
    ///     </c>
    /// </summary>
    public class AwaitExpr : AtomExpr {
        private Token? kwAwait;

        public Token? KwAwait {
            get => kwAwait;
            set => kwAwait = BindNullable(value);
        }

        private Expr val = null!;

        public Expr Value {
            get => val;
            set => val = Bind(value);
        }

        public override TypeName ValueType => Value.ValueType;

        public AwaitExpr(Node parent) : base(parent) { }

        public AwaitExpr Parse() {
            KwAwait = Stream.Eat(KeywordAwait);
            Value   = Multiple<Expr>.ParseGenerally(this);
            return this;
        }
    }
}
